ecosystem/sep-0035.md

```diff
@@ -0,0 +1,52 @@
+## Preamble
+
+```
+SEP: 0035
+Title: Operation IDs
+Author: Isaiah Turner (@Isaiah-Turner), Debnil Sur (@debnil), Scott Fleckenstein (@nullstyle)
+Track: Standard
+Status: Draft
+Created: 2020-08-26
+Discussion: https://groups.google.com/g/stellar-dev/c/vCgQhmox32Q
+```
+
+## Simple Summary
+This protocol defines an ID scheme for Stellar operations.
+
+## Motivation
+There is no universal method to generate IDs for Stellar operations. Horizon generates IDs, but that [implementation](https://github.com/stellar/go/blob/master/services/horizon/internal/toid/main.go) is internal and changeable. In many ways, this is already a de facto ecosystem standard: current prominent projects like Stellar Expert already utilize this scheme internally, and recent upgrades in Horizon's ingestion system will motivate other data-intensive applications. An example is SDF's own [Stellar ETL project](https://github.com/stellar/stellar-etl). But these applications cannot assign IDs in a consistent manner with Horizon's implementation. An ecosystem standard will unlock this consistency.
+
+## Abstract
+This protocol defines an assignment method for IDs for operations in historical ledger data, representing the total order of operations on the Stellar network. The ledger sequence number, transaction application order, and operation index are packed into a single ID value.
+
+## Specification
+An ID is a single 64-bit signed integer. The sign guarantees SQL compatibility, since some dialects do not support unsigned integers.
+In order to create an id, three pieces of information are required. All three should be expressed in network byte order, which is big-endian.
+
+```
+0                   1                   2                   3
+0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
++-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+|                    Ledger Sequence Number                     |
++-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+|     Transaction Application Order     |       Op Index        |
++-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+```
+
+1. **Ledger Sequence**: this takes up the last 32 bits of the id. The maximum supported number of ledgers is 2,147,483,647.
+2. **Transaction Application Order**: the order that the transaction was applied within the ledger where it was validated. This takes up the middle 20 bits of the id. The index starts at 1. The maximum supported number of transactions per operation is 1,048,575.
+3. **Operation Index**: the index of the operation within its parent transaction. This takes up the first 12 bits of the id. The index starts at 1. The maximum supported number of operations per transaction is 4095.
```

Comment:
following up on the first comment. Is this the first 12 bits or the last 12 bits?